### PR TITLE
Refactor construction API

### PR DESCRIPTION
**Pull request by andreibancioiu**
- Update / refactor Construction API to better accommodate  upcoming support for custom currencies (ESDT).
- In rosetta specification `v1.4.14`, `maxFee` and `feeMultiplier` do not exist anymore. See: https://github.com/coinbase/rosetta-specifications/pull/110, https://github.com/coinbase/rosetta-sdk-go/pull/439

### CODE REVIEW

File: server/services/common.go [hunk 1]
```diff
@@ -1,3 +1,54 @@
...
+	err = json.Unmarshal(data, value)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isZeroAmount(amount string) bool {
+	return amount == "" || amount == "0" || amount == "-0"
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply server/services/common.go:41]: Right, added a more robust solution - for integers, though (decimal numbers do not need to be handled - at least, not yet).

File: server/services/constructionMetadata.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		return errors.New("missing metadata: 'receiver'")
+	}
+	if metadata.GasLimit == 0 {
+		return errors.New("missing metadata: 'gasLimit'")
+	}
+	if metadata.GasPrice == 0 {
+		return errors.New("missing metadata: 'gasPrice'")
+	}
+	if metadata.Version != 1 {
+		return errors.New("bad metadata: unexpected 'version'")
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply server/services/constructionMetadata.go]: Indeed, fixed.

File: server/services/constructionMetadata.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	return txJson, nil
+}
+
+func (metadata *constructionMetadata) toTransaction() (*data.Transaction, error) {
+	err := metadata.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	tx := &data.Transaction{}
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply server/services/constructionMetadata.go:54]: Fixed.

File: server/services/constructionMetadata.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func (metadata *constructionMetadata) validate() error {
+	if len(metadata.Sender) == 0 {
+		return errors.New("missing metadata: 'sender'")
+	}
+	if len(metadata.Receiver) == 0 {
+		return errors.New("missing metadata: 'receiver'")
+	}
+	if metadata.GasLimit == 0 {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply server/services/constructionMetadata.go:75]: Yes, out of scope (at least for now). But good point nonetheless.

File: server/services/constructionService.go [hunk 5]
```diff
@@ -35,220 +33,169 @@ func (service *constructionService) ConstructionPreprocess(
...
-	suggestedFee, gasPrice, gasLimit, err := service.computeSuggestedFeeAndGas(txType, request.Options, networkConfig)
+	computedData := service.computeData(requestOptions)
+
+	fee, gasLimit, gasPrice, errTyped := service.computeFeeComponents(requestOptions, computedData)
 	if err != nil {
-		return nil, err
+		return nil, errTyped
+	}
+
+	metadata := &constructionMetadata{}
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply server/services/constructionService.go]: Fixed.

**Review by bogdan-rosianu — APPROVED** [review]

File: server/services/common.go [hunk 6]
```diff
@@ -38,7 +38,16 @@ func fromObjectsMap(obj objectsMap, value any) error {
...
 
 func isZeroAmount(amount string) bool {
-	return amount == "" || amount == "0" || amount == "-0"
+	if amount == "" {
+		return true
+	}
+
+	value, ok := big.NewInt(0).SetString(amount, 10)
+	if ok {
+		return value.Sign() == 0
```
Comment (iulianpascalau) [inline server/services/common.go:47]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**